### PR TITLE
Messages and Remote Func refactoring

### PR DESCRIPTION
**Pull request by srtaalej**
condensed app_mention, message_sent, and dm_sent callbacks
api_keys_dict for remote functions (**review request**: @hello-ashleyintech @zimeg @WilliamBergamin should each user have to input their api key in workflow builder or should Developer's key be used)

### CODE REVIEW

**Comment by WilliamBergamin** [comment]
Good question 💯  it depends on the use case of the app, it would be nice if each user could use their own API key but using the Developer's key might be sufficient for what the apps needs to do

From what I understand this sample is to be used as an example for developers, designing it to work with the Developer's key might be sufficient.

We can leave it up to the developer writing there own version of this project to implement support for individual user key

Since this is your project let me know what you think, either path can work imo

File: ai/handle_ai_selection.py [hunk 1]
```diff
@@ -3,11 +3,21 @@
 
 api=None
 current_api = None
+
 api_classes = {
         "openai": OpenAI_API,
         "anthropic": AnthropicAPI
         #your_llm_name: YourLLMClass
```
Comment (hello-ashleyintech) [inline ai/handle_ai_selection.py:10]: is this comment being left in or should we remove it completely? if we keep it in, could we add a comment block above `api_classes` describing the process to add a new LLM?

File: listeners/events/message_sent.py [hunk 2]
```diff
@@ -1,14 +1,101 @@
...
+            conversation_context= parse_messages(messages)
+
+            system_content = "This is a private DM between you and {user_id}. You are their helpful assistant."
+
+            response = generate_response(text, conversation_context, system_content)
+            say(text=response, thread_ts=thread_ts)
+
+
+
+        elif thread_ts: #only listening to non-mention messages that are in threads with the bot.
```
Comment (hello-ashleyintech) [inline listeners/events/message_sent.py:85]: can we add this comment in the code block on a new line?

File: listeners/events/message_sent.py [hunk 3]
```diff
@@ -1,14 +1,101 @@
...
+            thread_ts=event['ts']
+
+        conversation_context=parse_messages(messages)
+
+        if text:
+            response = generate_response(text, conversation_context)
+            say(text=response, thread_ts=thread_ts)
+
+        else:
+            response = "Hi there! You didn't provide a message with your mention. Mention me again in this thread so that I can help you out!"
```
Comment (hello-ashleyintech) [inline listeners/events/message_sent.py:49]: wondering if for these preset prompts, we should store them all in one centralized file and then reference those from that file?

**Review by hello-ashleyintech — COMMENTED** [review]
left a few comments 🙌 it's looking great so far! 😄 looks like there might be some linting/formatting issues also - let me know if you need help resolving any of them!

File: ai/handle_ai_selection.py (hunk 1 above)
Comment (srtaalej) [reply ai/handle_ai_selection.py:10]: i moved the instances dict into api_classes.py, i think it doesnt need the comment block if its there

File: listeners/events/message_sent.py (hunk 2 above)
Comment (srtaalej) [reply listeners/events/message_sent.py:85]: comment removed, that was just me keeping track of what things are doing 😅

File: listeners/events/message_sent.py (hunk 3 above)
Comment (srtaalej) [reply listeners/events/message_sent.py:49]: good idea!

**Pull request merged**